### PR TITLE
Second tank

### PR DESCRIPTION
**Pull request by tymochko**
Please review our current progress for this week: 
- after player invite creates new game in DB
- redirects each of players (invitor and invited) to game page with data from DB
- instantiate Game instance for each player and draw tanks in each player's window

### CODE REVIEW

File: src/client/languages/languages.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    UserNamePlaceholder: "Ім'я користувача",
+    UserAgePlaceholder: 'Вік',
+    UserOldPasswordPlaceholder: 'Старий пароль',
+    UserNewPasswordPlaceholder: 'Новий пароль',
+    UserConfirmNewPasswordPlaceholder: 'Підтвердіть новий пароль',
+    ErrorUserName1: "Ім'я користувача повинно складатися лише з букв англійського алфавіту і цифр ",
+    ErrorUserName2: " Ім'я користувача повинно складатися принаймні з 5 символів",
+    ErrorUserName3: 'Username cannot be longer than 15 characters',
+    ErrorUserAge1: 'Вам повинно бути не менше 18 років',
+    ErrorUserAge2: 'Ви не можете бути старше 100 років ',
```
Comment (olesiabrkt) [inline src/client/languages/languages.js]: =(((

File: src/client/models/externalFunctions.js [hunk 2]
```diff
@@ -35,3 +33,43 @@ export function fillBackground(ctx, pattern) {
...
+        thisWindowPlayerTurn = gameInst.player1.turn;
+        thisWindowPlayerId = gameInst.player1.id;
+    } else {
+        thisWindowPlayer = 'player2';
+        siblingWindowPlayer= 'player1';
+        thisWindowPlayerTurn = gameInst.player2.turn;
+        thisWindowPlayerId = gameInst.player2.id;
+    }
+
+    if (thisWindowPlayerTurn === true) {
```
Comment (olesiabrkt) [inline src/client/models/externalFunctions.js]: `if (thisWindowPlayerTurn) {` - the same

File: src/client/models/externalFunctions.js [hunk 3]
```diff
@@ -35,3 +33,43 @@ export function fillBackground(ctx, pattern) {
...
     ctx.fill();
 }
+
+export const drawTanks = (callback, tank1, tank2, tankImage, weaponImage) => {
+    canvasModel.getTank().ctx.clearRect(0, 0, WIDTH, HEIGHT);
+    callback(tank1, tankImage, weaponImage);
+    callback(tank2, tankImage, weaponImage);
+};
+
+export const allowTurn = (gameInst) => {
```
Comment (olesiabrkt) [inline src/client/models/externalFunctions.js]: the name of the function says like it should return boolean for me...

File: src/client/models/externalFunctions.js [hunk 4]
```diff
@@ -35,3 +33,43 @@ export function fillBackground(ctx, pattern) {
...
+        document.getElementById(thisWindowPlayer).style.color = 'red';
+        return thisWindowPlayerId;
+    }
+
+    document.getElementById(siblingWindowPlayer).style.color = 'red';
+    return null;
+};
+
+export const changeTurn = (playerTurn) => {
+    if (playerTurn) {
```
Comment (olesiabrkt) [inline src/client/models/externalFunctions.js]: haha, really? what for? just use `!playerTurn `

File: src/client/models/externalVariables.js [hunk 5]
```diff
@@ -1,8 +1,8 @@
 export const WIDTH = 800;
 export const HEIGHT = 500;
-export const TANKWIDTH = 70;
-export const TANKHEIGHT = 30;
-export const WEAPONWIDTH = 35;
-export const WEAPONHEIGHT = 20;
+export const TANK_WIDTH = 70;
+export const TANK_HEIGHT = 30;
+export const WEAPON_WIDTH = 35;
+export const WEAPON_HEIGHT = 20;
```
Comment (olesiabrkt) [inline src/client/models/externalVariables.js]: nice

File: src/client/models/gameModel.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        this.gameStatus = gameStatus;
+    }
+}
+
+module.exports.Game = Game;
+
+
+
+
+// gameObj;
```
Comment (olesiabrkt) [inline src/client/models/gameModel.js]: should be removed

File: src/client/models/navPanel.js [hunk 7]
```diff
@@ -1,58 +1,93 @@
...
+
+    socket.on('powerChangeAns', (data) => {
+        tankPower(data.tankPowerChange, data.power);
+    });
+
     getId('morePower').onclick = () => {
-        power++;
-        getId('power').innerHTML = power;
         power = parseInt(getId('power').innerHTML);
+        power += 5;
```
Comment (olesiabrkt) [inline src/client/models/navPanel.js]: what 5 means?

File: src/client/models/navPanel.js [hunk 8]
```diff
@@ -1,58 +1,93 @@
...
+        const thisPlayerId = localStorage.getItem('playerId');
+        gameInst.gameStatus = false;
+
+        if (gameInst.player1.id === thisPlayerId) {
+            gameInst.player1.life = 0;
+        } else {
+            gameInst.player2.life = 0;
+        }
+
+        socket.emit('end-game', gameInst);
```
Comment (olesiabrkt) [inline src/client/models/navPanel.js]: I hope server emits end game for the 2nd user, right?

File: src/client/models/shotTrajectory.js [hunk 9]
```diff
@@ -1,61 +1,78 @@
...
+    gameData.player1.tank = tank1;
+    gameData.player2.tank = tank2;
+
+};
+
 let originalPoints,
     lastTime,
-    dt2 = 0,
+    deltaT = 0,
+    deltaT_tank1 = 0,
```
Comment (olesiabrkt) [inline src/client/models/shotTrajectory.js]: do not use underscore with camelcase please

File: src/client/models/shotTrajectory.js [hunk 10]
```diff
@@ -1,61 +1,78 @@
...
-    lastFire = Date.now();
+
     shotStart();
 };
 
-module.exports.makeShot = makeShot;
-const shotStart = () => {
-
+const shotStart = (check_deltaT = true) => {
+	if(!check_deltaT) deltaT = deltaT_tank1;
```
Comment (olesiabrkt) [inline src/client/models/shotTrajectory.js]: forgot brackets

File: src/client/models/shotTrajectory.js [hunk 11]
```diff
@@ -64,39 +81,93 @@ const drawBullet = () => {
...
+    deltaT += 4*dt;
 
-    dt2 += 4*dt;
     // creating path for bullet and originalPoints
-    var bull = new paper.Path.Rectangle(bullet.pos[0],bullet.pos[1], 45, 7);
+    var bull = new paper.Path.Rectangle(new paper.Point(bullet.pos[0], bullet.pos[1]), new paper.Size(45, 7));
     //check angle for accuracy of point
-    bull.rotate(-bullet.imgInf.currAngle);
+    bull.rotate(-bullet.imgInf.currAngle * 180 / Math.PI);
+    bull.strokeColor = '#ff0000';
```
Comment (olesiabrkt) [inline src/client/models/shotTrajectory.js]: why have we to highlight it with red?

File: src/client/models/shotTrajectory.js [hunk 12]
```diff
@@ -64,39 +81,93 @@ const drawBullet = () => {
...
+    );
     for(let i = 1; i < originalPoints.length; i++) {
-        groundPath.add(new paper.Point(originalPoints[i][0], originalPoints[i][1]))
+        groundPath.add(new paper.Point(originalPoints[i][0], originalPoints[i][1]));
     }
+    groundPath.strokeColor = '#000';
+
+    var tankSize = new paper.Size(80,30);
+    var user1 = new paper.Path.Rectangle(new paper.Point(player1.data.tankX - tankSize.width / 2, player1.data.tankY - tankSize.height / 2), tankSize);
+    user1.strokeColor = '#00ff00';
```
Comment (olesiabrkt) [inline src/client/models/shotTrajectory.js]: what these colors for?

File: src/client/models/shotTrajectory.js [hunk 13]
```diff
@@ -64,39 +81,93 @@ const drawBullet = () => {
...
 
+    user1.rotate(-player1.data.tankAngle * 180 / Math.PI);
+    user2.rotate(-player2.data.tankAngle * 180 / Math.PI);
+
+
+    paper.view.draw();
     // check if intersect the original points
+    var intersectPlayer1 = bull.getIntersections(user2);
+    if(intersectPlayer1.length > 0) {
+        console.log('boom2');
```
Comment (olesiabrkt) [inline src/client/models/shotTrajectory.js]: ha! :)) remove it

File: src/client/models/shotTrajectory.js [hunk 14]
```diff
@@ -107,40 +178,35 @@ const generateExplosion = (dt) => {
...
+
+        gameData.points = calculatedGroundPoints;
+        
+        gameData.player1.turn = changeTurn(gameData.player1.turn);
+        gameData.player2.turn = changeTurn(gameData.player2.turn);
+        
+        socket.emit('update-data', gameData);
 
-        ground.setGround(calculatedGroundPoints);
+        // ground.setGround(calculatedGroundPoints);
```
Comment (olesiabrkt) [inline src/client/models/shotTrajectory.js]: just run lint :) I will not be pointing every single item like this.

File: src/client/models/tankShot.js [hunk 15]
```diff
@@ -1,120 +1,267 @@
...
+            tank1,
+            tank2
+        });
+    };
 
     const doKeyDown = (evt) => {
-        let now = new Date().getTime();
+        const now = new Date().getTime();
+
+        if ((now - lastTimeTankMoved > 800) && (allowTurn(gameInst) === localStorage.getItem('playerId'))) {
```
Comment (mnkami) [inline src/client/models/tankShot.js]: move 800 to some const

File: src/client/models/tankShot.js [hunk 16]
```diff
@@ -1,120 +1,267 @@
...
                 case 13: /*ENTER*/
-                    makeShot(canvasModel.getBullet().ctx, backCanvas, backCtx, tank.getCoord().tankX, tank.getCoord().tankY, tank.getTankAngle(), socket);
-                break;
+                    bulletToMove();
+                    tank1.tankY = findLinePoints(tank1.tankX);
+                    tank2.tankY = findLinePoints(tank2.tankX);
+                    drawTanks(drawTank, tank1, tank2, tankImage, weaponImage);
+                    break;
 
+                default:
```
Comment (mnkami) [inline src/client/models/tankShot.js]: you could remove default as you are not performing anything

File: src/client/models/tankShot.js [hunk 17]
```diff
@@ -1,120 +1,267 @@
...
-    window.addEventListener('keydown',doKeyDown,true);
 
-/* ======   Navigation ======== */
+    document.addEventListener('keydown', moveWeaponKeyDown, true);
+    window.addEventListener('keydown', doKeyDown, true);
 
-    navPanel(angle, tankX, tankY, weaponAngle);
+    socket.on('outputPosTank', (data) => {
+        tankMove(data.direction, data.tankMoves, data.tank1, data.tank2, tankImage, weaponImage, socket);
+        console.log(allowTurn(gameInst), 'allowTurn');
```
Comment (mnkami) [inline src/client/models/tankShot.js]: still in progress?

File: src/client/modules/dashboard/index.js [hunk 18]
```diff
@@ -1,28 +1,77 @@
...
+    });
+
+    socket.on('start-game', function(foundGame) {
+    //    socket.emit('points', {points:foundGame.originalPoints});
+        localStorage.setItem('playerId', foundGame.playerId);
+        window.location = `/game?id=${foundGame.gameId}`;
+    });
+
+    socket.on('invite-rejected', function(data) {
+        alert('Your invitation was rejected.');
```
Comment (mnkami) [inline src/client/modules/dashboard/index.js:53]: should be a modal already

File: src/client/modules/dashboard/requestToPlay.html [hunk 19]
```diff
@@ -0,0 +1,9 @@
+<div class="modal-header">
+</div>
+<div class="modal-body">
+    <h3 class="modal-title">Wanna play with  {{data.senderUsername}} ?</h3>
```
Comment (mnkami) [inline src/client/modules/dashboard/requestToPlay.html:4]: @MykhailoDzedzei translate

File: src/client/modules/game/gameService.js [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+            });
+            return deferred.promise;
+        },
+
+        getInitGameData(socket, $q, gameId) {
+
+            const deferred = $q.defer();
+
+            socket.on('get-game-data', (gameData) => {
+                console.log('2');
```
Comment (mnkami) [inline src/client/modules/game/gameService.js]: remove

File: src/client/modules/profile/ManageProfileController.js [hunk 21]
```diff
@@ -26,32 +26,52 @@ export class ManageProfileController {
 
         $scope.selectedImg = "api/users/profile/getImage/userAvatar" + ($scope.getSalt)();
 
-        function savingMsg () {
-            toastr.success('Your changes are saved!', 'Message', {
-                closeButton: true,
-                closeHtml: '<button>&times;</button>'
-            });
+        function savingMsg() {
+            $translate('savingMsg')
```
Comment (mnkami) [inline src/client/modules/profile/ManageProfileController.js:30]: use Promise.all

File: src/client/modules/profile/ManageProfileController.js [hunk 22]
```diff
@@ -26,32 +26,52 @@ export class ManageProfileController {
...
+                });
+            $window.location.reload();
         }
 
-        function avatarMsg () {
-            toastr.warning('Do not forget to save changes!!', 'Message', {
-                closeButton: true,
-                closeHtml: '<button>&times;</button>'
+        function avatarMsg() {
+            $translate('avatarMsg')
```
Comment (mnkami) [inline src/client/modules/profile/ManageProfileController.js:45]: Promise.all

File: src/server/api/users/usersController.js [hunk 23]
```diff
@@ -280,7 +286,11 @@ const getUserImage = function (req, res) {
             return res.status(403).send();
         }
         userImage = foundUser.userImg;
+        if(userImage.uploadedImg === null) {
+            return res.status(403).send();
+        }
         if (userImage.uploadedImg) {
+
             userDir = __dirname + '/../../static/usersInfo/' + userId + '/' + userImage.image;
```
Comment (mnkami) [inline src/server/api/users/usersController.js]: @MykhailoDzedzei you always forget about these lines!!! :)

File: src/server/socket/chat.js [hunk 24]
```diff
@@ -0,0 +1,37 @@
+let mongoose = require('mongoose');
```
Comment (mnkami) [inline src/server/socket/chat.js]: fix the code style

File: src/client/models/externalFunctions.js [hunk 25]
```diff
@@ -35,3 +33,43 @@ export function fillBackground(ctx, pattern) {
...
+    if (thisWindowPlayerTurn === true) {
+        document.getElementById(thisWindowPlayer).style.color = 'red';
+        return thisWindowPlayerId;
+    }
+
+    document.getElementById(siblingWindowPlayer).style.color = 'red';
+    return null;
+};
+
+export const changeTurn = (playerTurn) => {
```
Comment (mnkami) [inline src/client/models/externalFunctions.js]: just `!playerTurn`

File: src/client/models/tankShot.js (hunk 17 above)
Comment (tymochko) [inline src/client/models/tankShot.js]: Yes, sorry. We decided today is our final day for implementation functionality.

File: src/server/socket/chat.js (hunk 24 above)
Comment (tymochko) [inline src/server/socket/chat.js]: sorry, what do you mean?

File: src/client/models/externalFunctions.js (hunk 25 above)
Comment (tymochko) [inline src/client/models/externalFunctions.js]: thanks. That's simple.

File: src/client/models/tankShot.js (hunk 16 above)
Comment (kachanovskyi) [inline src/client/models/tankShot.js]: sorry, ESLint wanted to have default case, that's why I wrote it.

**Pull request merged**